# Анализ базы данных сервиса по чтению книг по подписке

**Цель исследования**

Провести анализ базы данных для формулирования ценностного предложения нового продукта.

**Описание данных**

Таблица `books`. Содержит данные о книгах:
- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

Таблица `authors`. Содержит данные об авторах:
- `author_id` — идентификатор автора;
- `author` — имя автора.

Таблица `publishers`. Содержит данные об издательствах:
- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

 Таблица `ratings`. Содержит данные о пользовательских оценках книг:
- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

Таблица `reviews`. Содержит данные о пользовательских обзорах на книги:
- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Импорт-библиотек" data-toc-modified-id="Импорт-библиотек-0"><span class="toc-item-num">0&nbsp;&nbsp;</span>Импорт библиотек</a></span></li><li><span><a href="#Подключение-к-базе-данных-и-ознакомление-с-данными" data-toc-modified-id="Подключение-к-базе-данных-и-ознакомление-с-данными-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подключение к базе данных и ознакомление с данными</a></span></li><li><span><a href="#Исследовательский-анализ-данных" data-toc-modified-id="Исследовательский-анализ-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследовательский анализ данных</a></span><ul class="toc-item"><li><span><a href="#Задание-№1" data-toc-modified-id="Задание-№1-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Задание №1</a></span></li><li><span><a href="#Задание-№2" data-toc-modified-id="Задание-№2-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Задание №2</a></span></li><li><span><a href="#Задание-№3" data-toc-modified-id="Задание-№3-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Задание №3</a></span></li><li><span><a href="#Задание-№4" data-toc-modified-id="Задание-№4-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Задание №4</a></span></li><li><span><a href="#Задание-№5" data-toc-modified-id="Задание-№5-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Задание №5</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

## Импорт библиотек

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

## Подключение к базе данных и ознакомление с данными

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})
con = engine.connect()

Для ознакомления выведем первые пять строк каждой таблицы.

In [8]:
query = '''
SELECT * 
FROM books 
LIMIT 5
'''
pd.io.sql.read_sql(sql=text(query), con=con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [9]:
query = '''
SELECT * 
FROM authors 
LIMIT 5
'''
pd.io.sql.read_sql(sql=text(query), con=con)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [10]:
query = '''
SELECT * 
FROM publishers 
LIMIT 5
'''
pd.io.sql.read_sql(sql=text(query), con=con)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [12]:
query = '''
SELECT * 
FROM ratings 
LIMIT 5
'''
pd.io.sql.read_sql(sql=text(query), con=con)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [13]:
query = '''
SELECT * 
FROM reviews 
LIMIT 5
'''
pd.io.sql.read_sql(sql=text(query), con=con)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

## Исследовательский анализ данных

### Задание №1

Посчитайте, сколько книг вышло после 1 января 2000 года.

In [16]:
query = '''
SELECT COUNT(book_id) 
FROM books 
WHERE CAST(publication_date AS date) > '2000-01-01'
'''
pd.io.sql.read_sql(sql=text(query), con=con)

count
0    819

После 1 января 2000 года вышло **819** книг.

### Задание №2

Для каждой книги посчитайте количество обзоров и среднюю оценку.

In [24]:
query = '''
SELECT b.book_id,
       b.title,
       COUNT(DISTINCT rev.review_id) AS cnt_reviews,
       ROUND(AVG(rat.rating), 2) AS avg_rating 
FROM books AS b
LEFT JOIN ratings AS rat on b.book_id=rat.book_id
LEFT JOIN reviews AS rev on b.book_id=rev.book_id
GROUP BY b.book_id
ORDER BY cnt_reviews DESC
'''
pd.io.sql.read_sql(sql=text(query), con=con)

book_id                                              title  cnt_reviews  \
0        948                            Twilight (Twilight  #1)            7   
1        963                                Water for Elephants            6   
2        734                                   The Glass Castle            6   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...            6   
4        695  The Curious Incident of the Dog in the Night-Time            6   
..       ...                                                ...          ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel            0   
996      808                            The Natural Way to Draw            0   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites            0   
998      221                          Essential Tales and Poems            0   
999      191  Disney's Beauty and the Beast (A Little Golden...            0   

     avg_rating  
0          3.66  
1          3.98  
2          4.21  
3          4.41  
4          4.08  
..          ...  
995        3.67  
996        3.00  
997        5.00  
998        4.00  
999        4.00  

[1000 rows x 4 columns]

Для каждой книги мы посчитали количество обзоров и среднюю оценку. Больше всего обзоров написано на книгу **Twilight (Twilight #1)**, 7 обзоров, её средний рейтинг при этом составляет 3,66 из 5.

### Задание №3

Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры.

In [30]:
query = '''
SELECT p.publisher_id,
       p.publisher,
       COUNT(b.book_id) as cnt_books
FROM publishers AS p
LEFT JOIN books as b on p.publisher_id=b.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher_id
ORDER BY cnt_books DESC
LIMIT 1
'''
pd.io.sql.read_sql(sql=text(query), con=con)

publisher_id      publisher  cnt_books
0           212  Penguin Books         42

Наибольшее число книг толще 50 страниц выпустило издательство **Penguin Books**.

### Задание №4

Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками.

In [36]:
query = '''
SELECT a.author,
       ROUND(AVG(r.rating), 2) AS avg_rating
FROM books AS b
INNER JOIN authors AS a on b.author_id=a.author_id
INNER JOIN ratings AS r on b.book_id=r.book_id
WHERE b.book_id in (
                    SELECT book_id
                    FROM ratings
                    GROUP BY book_id
                    HAVING COUNT(book_id) >= 50
                    )
GROUP BY a.author
ORDER BY avg_rating DESC
LIMIT 1
'''
pd.io.sql.read_sql(sql=text(query), con=con)

author  avg_rating
0  J.K. Rowling/Mary GrandPré        4.29

Автор с самой высокой средней оценкой книг, учитывая только книги, которые имеют 50 и более оценок - **J.K. Rowling/Mary GrandPré**. Средняя оценка её книг составляет **4,29**.

### Задание №5

Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [39]:
query = '''
WITH users_rev_num AS (
SELECT rev.username,
       COUNT(rev.review_id) AS cnt_reviews
FROM reviews as rev
WHERE rev.username IN (SELECT username
                       FROM ratings
                       GROUP BY username
                       HAVING COUNT(rating_id) > 48
                       )
GROUP BY rev.username
)
SELECT AVG(cnt_reviews) AS avg_reviews
FROM users_rev_num
'''
pd.io.sql.read_sql(sql=text(query), con=con)

avg_reviews
0         24.0

В среднем пользователи, которые поставили более 48 оценок, оставляют **24** отзыва.

## Общий вывод

Мы провели анализ базы данных сервиса чтения книг по подписке. В ходе исследования мы пришли к следующим выводам:
- после 1 января 2000 года вышло **819** книг
- больше всего обзоров написано на книгу Twilight (Twilight #1), всего **7** обзоров
- наибольшее число книг толще 50 страниц выпустило издательство **Penguin Books**
- автор с самой высокой средней оценкой книг - **J.K. Rowling/Mary GrandPré**. Средняя оценка её книг составляет **4,29**
- в среднем пользователи, которые поставили более 48 оценок, оставляют **24** отзыва